### PR TITLE
DirectILL input numors in the SampleLogs

### LINKED ISSUE #31883: DirectILL reduction: acess to input and corrections numors
A user of the DirectILL reduction chain has reported that the access to the numors that are used to build workspaces in the workflow is not practical and readable. While it is possible to extract this information from a workspace's history, there it contains the full paths and its readability is diminished.

A solution needs to provide information about the direct inputs used to build a workspace in `DirectILLCollectData` as a first step, and preferably also include numors of workspaces used to perform corrections.

### Platforms affected
all

### PR DESCRIPTION
**Pull request by darominski**
This PR partially addresses the issue of retrieving the ids (numors) of files that were used to create final output data in the Direct ILL reduction workflow. It adds a new sample log: 'run_list' to two NeXus loaders handling ILL direct geometry instruments: `LoadILLTOF2` and `LoadILLDiffraction` algorithms to ease the extraction of numors used to load the data. The information containing numors of correction workspaces will be added in a separate reduction algorithm, currently in development.

**To test:**

1. Load one or more data files using either directly `LoadAndMerge` or indirectly through `DirectILLCollectData`, for example using PANTHER's unit test data under: Data/UnitTest/ILL/PANTHER
2. Observe if the SampleLog 'run_list' is created and the value contains all the file names in your input.

Partially fixes #31883 . 

Release note added.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataHandling/src/LoadILLDiffraction.cpp [hunk 1]
```diff
@@ -258,6 +258,7 @@ void LoadILLDiffraction::loadMetaData() {
     m_loadHelper.addNexusFieldsToWsRun(nxHandle, m_outWorkspace->mutableRun());
     NXclose(&nxHandle);
   }
+  mutableRun.addProperty("run_list", std::string(mutableRun.getProperty("run_number")->value()));
```
Comment (gvardany) [inline Framework/DataHandling/src/LoadILLDiffraction.cpp:261]: `getPropertyValue()` should return you a string directly.

File: Framework/DataHandling/src/LoadILLTOF2.cpp [hunk 2]
```diff
@@ -273,6 +273,7 @@ void LoadILLTOF2::addAllNexusFieldsAsProperties(const std::string &filename) {
   }
   m_loader.addNexusFieldsToWsRun(nxfileID, runDetails);
   NXclose(&nxfileID);
+  runDetails.addProperty("run_list", std::string(runDetails.getProperty("run_number")->value()));
```
Comment (gvardany) [inline Framework/DataHandling/src/LoadILLTOF2.cpp:276]: same here

File: Framework/DataHandling/test/LoadILLDiffractionTest.h [hunk 3]
```diff
@@ -686,6 +693,7 @@ class LoadILLDiffractionTest : public CxxTest::TestSuite {
 
     TS_ASSERT_DELTA(ei, 5.12, 0.01)
     TS_ASSERT_EQUALS(outputWS->run().getProperty("Detector.calibration_file")->value(), "none")
+    TS_ASSERT_EQUALS(runList->value(), std::string("104"))
```
Comment (gvardany) [inline Framework/DataHandling/test/LoadILLDiffractionTest.h:696]: Do you need `std::string`?

File: Framework/DataHandling/test/LoadILLTOF2Test.h [hunk 4]
```diff
@@ -133,8 +136,12 @@ class LoadILLTOF2Test : public CxxTest::TestSuite {
...
     const size_t monitorCount = 1;
     const bool convertToTOF = true;
-    loadDataFile("ILL/IN5/104007.nxs", histogramCount, monitorCount, channelCount, tofDelay, tofChannelWidth,
-                 convertToTOF);
+    MatrixWorkspace_sptr ws = loadDataFile("ILL/IN5/104007.nxs", histogramCount, monitorCount, channelCount, tofDelay,
+                                           tofChannelWidth, convertToTOF);
+    auto const run = ws->run();
+    TS_ASSERT(run.hasProperty("run_list"))
+    const auto runList = run.getLogData("run_list");
+    TS_ASSERT_EQUALS(runList->value(), std::string("104007"))
```
Comment (gvardany) [inline Framework/DataHandling/test/LoadILLTOF2Test.h:144]: same here

File: Framework/DataHandling/test/LoadILLTOF2Test.h [hunk 5]
```diff
@@ -148,8 +155,12 @@ class LoadILLTOF2Test : public CxxTest::TestSuite {
     MatrixWorkspace_sptr ws = loadDataFile("ILL/IN6/164192.nxs", histogramCount, monitorCount, channelCount, tofDelay,
                                            tofChannelWidth, convertToTOF);
 
-    const double pulseInterval = ws->run().getLogAsSingleValue("pulse_interval");
+    auto const run = ws->run();
+    const double pulseInterval = run.getLogAsSingleValue("pulse_interval");
     TS_ASSERT_DELTA(0.0060337892, pulseInterval, 1e-10)
+    TS_ASSERT(run.hasProperty("run_list"))
+    const auto runList = run.getLogData("run_list");
+    TS_ASSERT_EQUALS(runList->value(), std::string("164192"))
```
Comment (gvardany) [inline Framework/DataHandling/test/LoadILLTOF2Test.h:163]: same here

File: Framework/DataHandling/test/LoadILLTOF2Test.h [hunk 6]
```diff
@@ -211,8 +222,12 @@ class LoadILLTOF2Test : public CxxTest::TestSuite {
...
     const size_t monitorCount = 1;
     const bool convertToTOF = false;
-    loadDataFile("ILL/PANTHER/001723.nxs", histogramCount, monitorCount, channelCount, tofDelay, tofChannelWidth,
-                 convertToTOF);
+    MatrixWorkspace_sptr ws = loadDataFile("ILL/PANTHER/001723.nxs", histogramCount, monitorCount, channelCount,
+                                           tofDelay, tofChannelWidth, convertToTOF);
+    auto const run = ws->run();
+    TS_ASSERT(run.hasProperty("run_list"))
+    const auto runList = run.getLogData("run_list");
+    TS_ASSERT_EQUALS(runList->value(), std::string("1723"))
```
Comment (gvardany) [inline Framework/DataHandling/test/LoadILLTOF2Test.h:230]: same here

File: Framework/DataHandling/test/LoadILLTOF2Test.h [hunk 7]
```diff
@@ -235,8 +250,12 @@ class LoadILLTOF2Test : public CxxTest::TestSuite {
...
     const size_t monitorCount = 1;
     const bool convertToTOF = false;
-    loadDataFile("ILL/SHARP/000102.nxs", histogramCount, monitorCount, channelCount, tofDelay, tofChannelWidth,
-                 convertToTOF);
+    MatrixWorkspace_sptr ws = loadDataFile("ILL/SHARP/000102.nxs", histogramCount, monitorCount, channelCount, tofDelay,
+                                           tofChannelWidth, convertToTOF);
+    auto const run = ws->run();
+    TS_ASSERT(run.hasProperty("run_list"))
+    const auto runList = run.getLogData("run_list");
+    TS_ASSERT_EQUALS(runList->value(), std::string("102"))
```
Comment (gvardany) [inline Framework/DataHandling/test/LoadILLTOF2Test.h:258]: same here

**Review by gvardany — COMMENTED** [review]
Looks good, just a few minor comments:

File: Framework/DataHandling/src/LoadILLDiffraction.cpp (hunk 1 above)
Comment (darominski) [reply Framework/DataHandling/src/LoadILLDiffraction.cpp:261]: `getPropertyValue()` is not a member function of  `Run` nor `LogManager` classes, it is possible to extract property values as strings from algorithms this way, but not from the mutable run object here. I agree the current solution is a bit of a mouthful, and perhaps a better alternative would be to use `getPropertyValueAsType<std::string>`.

File: Framework/DataHandling/test/LoadILLDiffractionTest.h (hunk 3 above)
Comment (darominski) [reply Framework/DataHandling/test/LoadILLDiffractionTest.h]: A string literal would be sufficient, actually, you are correct.

**Comment by gvardany** [comment]
This is a little confusing now; run_list is originally of type int on each file, but then it becomes string when merging due to the `,` ? Or is the list also treated as int array.

**Comment by darominski** [comment]
> This is a little confusing now; run_list is originally of type int on each file, but then it becomes string when merging due to the `,` ? Or is the list also treated as int array.

I believe it is castable to string, but the reason it is tested against a string literal is due to getting its value using "value()", as in L128 of `LoadILLTOF2Test.h`, which by definition returns the value of the property as a string.

**Review by gvardany — DISMISSED** [review]
LGTM.

**Review by gvardany — APPROVED** [review]
Reapprove after merge with master.

**Review by DavidFair — COMMENTED** [review]
LGTM, I'm happy to merge this
Just one question, I double checked `LogManager.cpp::addProperty` and this will remove the property (if overwrite is set), or do nothing if overwrite is not set and a value exists.
Since this is labelled as a list I'm guessing the long-term intention is to hold multiple run numbers here. But this approach will only allow you to store a single value since were packing an int here instead of a std::vector or similar?

I may be misinterpreting how LogManager works though

**Comment by darominski** [comment]
> LGTM, I'm happy to merge this
> Just one question, I double checked `LogManager.cpp::addProperty` and this will remove the property (if overwrite is set), or do nothing if overwrite is not set and a value exists.
> Since this is labelled as a list I'm guessing the long-term intention is to hold multiple run numbers here. But this approach will only allow you to store a single value since were packing an int here instead of a std::vector or similar?
> 
> I may be misinterpreting how LogManager works though

In the loader we load only one file at a time, so this property needs to take only one int at the time of creation. Then, during reduction, `MergeRuns` turns it into a list of integers.

**Pull request merged**